cmd/state/autoupdate.go

```diff
@@ -0,0 +1,164 @@
+			logging.Error("Failed to store last update check: %s", errs.JoinMessage(err))
+		}
+	}()
+	
+	if !shouldRunAutoUpdate(args, cfg) {
+		return false, nil
+	}
+
+	// Check for available update
+	checker := updater.NewDefaultChecker(cfg)
+	up, err := checker.Check()
+	if err != nil {
+		return false, errs.Wrap(err, "Failed to check for update")
+	}
+	if up == nil {
+		logging.Debug("No update found")
+		return false, nil
+	}
+
+	out.Notice(output.Heading(locale.Tl("auto_update_title", "Auto Update")))
+	out.Notice(locale.Tr("auto_update_to_version", constants.Version, up.Version))
+
+	logging.Debug("Auto updating to %s", up.Version)
+
+	// Protect against multiple updates happening simultaneously
+	fileLock := flock.New(filepath.Join(cfg.ConfigPath(), "install.lock"))
+	lockSuccess, err := fileLock.TryLock()
+	if err != nil {
+		return false, errs.Wrap(err, "Could not create file lock required to install update")
+	}
+	if !lockSuccess {
+		logging.Debug("Another update is already in progress")
+		return false, nil
+	}
+	defer fileLock.Unlock()
+
+	targetDir := filepath.Dir(appinfo.StateApp().Exec())
+	err = up.InstallBlocking(targetDir)
+	if err != nil {
+		log := logging.Error
+		innerErr := errs.InnerError(err)
+		if os.IsPermission(innerErr) {
+			return false, locale.WrapInputError(err, "auto_update_permission_err", "", constants.DocumentationURL, errs.JoinMessage(err))
+		}
+		if errors.As(err, new(*lockfile.AlreadyLockedError)) {
+			log("Auto update failed because the update lock file is already in use")
+			return false, nil
+		}
+		return false, errs.Wrap(err, "Failed to install update")
+	}
+
+	out.Notice(locale.Tr("auto_update_relaunch"))
+	out.Notice("") // Ensure output doesn't stick to our messaging
+	code, err := relaunch(args)
+	if err != nil {
+		return true, errs.WrapExitCode(err, code)
+	}
+
+	return true, nil
+}
+
+func shouldRunAutoUpdate(args []string, cfg *config.Instance) bool {
+	switch {
+	// In a forward
+	case os.Getenv(constants.ForwardedStateEnvVarName) == "true":
+		logging.Debug("Not running auto updates because we're in a forward")
+		return false
+
+	// Forced enabled (breaks out of switch)
+	case os.Getenv(constants.TestAutoUpdateEnvVarName) == "true":
+		logging.Debug("Forcing auto update as it was forced by env var")
+		return true
+
+	// In unit test
+	case condition.InUnitTest():
+		logging.Debug("Not running auto updates in unit tests")
+		return false
+
+	// Already running manual update
+	case funk.Contains(args, "update"):
+		logging.Debug("Not running auto updates because 'update' in args")
+		return false
+
+	// Updates are disabled
+	case strings.ToLower(os.Getenv(constants.DisableUpdates)) == "true":
+		logging.Debug("Not running auto updates because updates are disabled by env var")
+		return false
+
+	// We're on CI
+	case (condition.OnCI()) && strings.ToLower(os.Getenv(constants.DisableUpdates)) != "false":
+		logging.Debug("Not running auto updates because we're on CI")
```

Comment:
Do we need the second part of the boolean expression (the one that checks for `DisableUpdates != "false"`.  We already know that it is not `true`, due to the check above...

Reply:
I'm pretty sure this is so that for certain CI tests we can explicitly say "do NOT disable updates". Since they are otherwise disabled by default.

I could be wrong though. This isn't new code, I just moved it into a switch.

Reply:
Ah! That makes sense.